### PR TITLE
Basic runtime support for dynamic providers

### PR DESCRIPTION
**Pull request by iwahbe**
This PR replaces the `.Replace` method with a sounder construct: `XParameterizeResetProvider`.

- To achieve a working result, it relaxes an assert in `ProposeNew`.
- To clarify how `XParameterizeResetProvider` works, a level of interface indirection was removed.

### CODE REVIEW

File: dynamic/main.go [hunk 1]
```diff
@@ -85,7 +80,14 @@ func initialSetup() (tfbridge.ProviderInfo, pfbridge.ProviderMetadata, func() er
 			}
 
 			tfServer = p
-			defaultInfo.P.(*proto.Provider).Replace(p)
+			if tfServer != nil {
+				info = providerInfo(ctx, tfServer)
+			}
+
+			err = pfbridge.XParameterizeResetProvider(ctx, info, metadata)
```
Comment (VenelinMartinov) [inline dynamic/main.go:87]: Is this the variable which contains the lambda? Is it passed as a reference?

File: dynamic/main.go [hunk 2]
```diff
@@ -85,7 +80,14 @@ func initialSetup() (tfbridge.ProviderInfo, pfbridge.ProviderMetadata, func() er
 			}
 
 			tfServer = p
-			defaultInfo.P.(*proto.Provider).Replace(p)
+			if tfServer != nil {
+				info = providerInfo(ctx, tfServer)
+			}
+
```
Comment (VenelinMartinov) [inline dynamic/main.go:86]: Do we support re-parametrising the provider without restarting it?

**Review by VenelinMartinov — APPROVED** [review]

File: dynamic/main.go (hunk 2 above)
Comment (iwahbe) [reply dynamic/main.go:86]: No, but there should probably be an explicit error there to signal that the parameterization was not successful.

File: dynamic/main.go (hunk 1 above)
Comment (iwahbe) [reply dynamic/main.go:87]: If your talking about the lambda introduced in provider.go:

https://github.com/pulumi/pulumi-terraform-bridge/blob/8c172bf6702ddd8403be87eb2b83fdc32bdebabb/pf/tfbridge/provider.go#L213-L222

Then `ctx` is the variable which contains the lambda. `XParameterizeResetProvider` unpacks the lambda from `ctx` and then calls it.

Comment (VenelinMartinov) [reply dynamic/main.go:87]: The `metadata` used on this line comes from line 51, as far as I can see, right?

The same `metadata` also contains this function which captures it.

It looks a bit odd and I was wondering if it's passed as a reference for this to work.

Comment (iwahbe) [reply dynamic/main.go:87]: Ah, I think I know what you mean. In Go, lambdas always have access to the location of variables in their scope. If your familiar with C++, then lambdas close over the lvalue of variables they reference.

**Pull request merged**